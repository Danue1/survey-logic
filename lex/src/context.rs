pub struct Context<'lex> {
    source: &'lex str,
    index: usize,
}

impl<'lex> Context<'lex> {
    pub const fn new(source: &'lex str) -> Self {
        Self { source, index: 0 }
    }

    pub const fn index(&self) -> usize {
        self.index
    }

    pub fn peek(&self) -> Option<char> {
        self.source.get(self.index..).and_then(|s| s.chars().next())
    }

    pub fn bump(&mut self) -> Option<char> {
        let c = self.peek()?;
        self.advance();
        Some(c)
    }

    #[inline]
    pub fn advance(&mut self) {
        self.index += 1;
    }

    #[inline]
    pub fn slice(&self, start: usize) -> String {
        self.source[start..self.index].to_owned()
    }
}
