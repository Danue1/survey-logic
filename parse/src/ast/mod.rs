mod condition;
mod literal;
mod question_parameter;
mod statement;

use crate::context::{Context, State};
use syntax_kind::SyntaxKind;

impl Context {
    pub fn program(&mut self) {
        self.start_node(SyntaxKind::PROGRAM_NODE);
        self.skip_whitespace();
        while !self.is_empty() {
            if self.statement() == State::Continue {
                self.unexpected_token();
            }
            self.skip_whitespace();
        }
        self.finish_node();
    }

    pub fn unexpected_token(&mut self) {
        self.start_node(SyntaxKind::ERROR);
        self.bump();
        self.finish_node();
    }
}

#[cfg(test)]
mod tests {
    #[test]
    fn statements() {
        let tokens = lex::lex("Q123 = 1 Q456 = 2");
        let mut context = super::Context::new(tokens);
        context.program();
        let node = context.finish();
        insta::assert_debug_snapshot!(node);
    }
}
