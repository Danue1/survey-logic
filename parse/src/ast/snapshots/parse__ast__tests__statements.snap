---
source: parse/src/ast/mod.rs
expression: node
---
PROGRAM_NODE@0..17
  STATEMENT_NODE@0..9
    COMPARISON_CONDITION_NODE@0..8
      BINARY_COMPARISON_NODE@0..8
        QUESTION_PARAMETER_NODE@0..4
          IDENTIFIER@0..1 "Q"
          NUMERIC_LITERAL_NODE@1..4
            NUMERIC@1..4 "123"
        WHITESPACE@4..5 " "
        EQUAL_OPERATOR_NODE@5..6
          EQUAL@5..6 "="
        WHITESPACE@6..7 " "
        NUMERIC_LITERAL_NODE@7..8
          NUMERIC@7..8 "1"
    WHITESPACE@8..9 " "
  STATEMENT_NODE@9..17
    COMPARISON_CONDITION_NODE@9..17
      BINARY_COMPARISON_NODE@9..17
        QUESTION_PARAMETER_NODE@9..13
          IDENTIFIER@9..10 "Q"
          NUMERIC_LITERAL_NODE@10..13
            NUMERIC@10..13 "456"
        WHITESPACE@13..14 " "
        EQUAL_OPERATOR_NODE@14..15
          EQUAL@14..15 "="
        WHITESPACE@15..16 " "
        NUMERIC_LITERAL_NODE@16..17
          NUMERIC@16..17 "2"

